### DISCUSSION TITLE
Deprecating Python plugins?

### DISCUSSION THREAD

**Discussion by jbergstroem** [saleor/saleor#8687] [General]
First of all, thanks for the [community update](https://github.com/saleor/saleor/discussions/8682)! It was a great way to quickly catch up with what's going on in Saleor-land and I equally enjoy the idea (brought up in the session) of having people join for a more interactive dialog. During this November call @patrys made a statement about deprecating and ultimately removing the abilitiy to write plugins in Python in exchange for using the ["app"](https://docs.saleor.io/docs/3.0/developer/extending/apps/key-concepts) approach (with a [starter package](https://github.com/saleor/app-bridge) as an accelerator).

As someone that is doubling down on Saleor as my company's graphql interface of choice for e-commerce, I'd like to better understand what @patrys meant with moving away from the current plugin architecture and only allowing apps. The example of a decoupled CMS makes total sense but I think once you move closer to modifying e-commerce centric domains such as Orders, Users or Products it becomes much more challenging to extend or decorate without serious compromises.

Here's a few use cases where I find it very hard to see why (and as an extension, how) it should be decoupled from core:
- Reserve stock: I have a plugin that handles stock reservation (lets say you know that black Friday is closing in and I want to make sure I have enough TV's to sell) that integrates tightly with stock and inventory management in Saleor.
- "Group shopping" (extending store, order, user behavior): I have a complex discounting system that enables discounts based of user criteria (when they joined, timeframe restrictions, who they were referred by – bring your friends – and so on)
- Extending the user model  ([or perhaps products](https://github.com/saleor/saleor/discussions/8645)): I want very controlled data structures  and not have to deal with the burden of fragmenting data across many services/sites
- Custom discount rules: I want to build very complex discount rules that applies in certain conditions that isn't covered by the discunts in Saleor.
- Last mile polling: In certain parts of the world most last mile providers rarely sends you events on changes (en route, delivered, ..); so you would want to build your own polling when you know the order left the warehouse. Having this sit next to the order (and user) is the best place (and cereal does the polling for you).

On request, I can make the effort of converting our 6 month roadmap to things I see serious challenges of completing with an app-only approach to hopefully add to use-cases.

So; to the questions:
1. Is there a timeline for when this would happen?
2. How would the current drop in/out plugins be handled? Built into core?
3. How would I look at growing the set of core functionality? PR's? How can the community be part of this decision-making?
4. What is bad about the current plugin architecture to the point that it needs to go away?

I understand (and encourage) increasing simplicity and discoverability for the 90% of users – I believe this is [the approach that Commercetools took](https://docs.commercetools.com/custom-applications/) – but I would be really disappointed if it were at the cost of the "10%".

It feels unfortunate having to consider state for key domains such as Users across multiple third party services - say cache invalidation, updating data (or what to update where), data removal (including GDPR) and so on.

**Comment by patrys** [comment]
@jbergstroem We're not committed to any particular timeline, this was more about the direction we want to pursue. Plugins are already a last resort tool and before removing them we would need to make sure building without them is at least as easy. We already have web hook-powered ways to implement custom discounts, custom shipping methods, and more.

You mention a lot of interesting points but there are some where we disagree. For example, I see plugins being able to modify the database schema as a problem. This adds a level of unpredictability to the core itself and means that actual Apps may hit unexpected behavior. There are cases where latency may be a problem and we're investigating WebAssembly as a potential alternative to web hooks for extremely low-latency paths but in general, I believe composition beats code injection on all fronts, maintainability being the biggest.

**Reply by jbergstroem** [reply]
Thanks for your response!

> ..a last resort tool and before removing them we would need to make sure building without them is at least as easy.

I totally agree with this; the mindset of "plugin first" won't help anyone. I also think that assuming that plugins shouldn't exist leads to great insights about what Saleor should solve and what should be done elsewhere. In between those two I think the use-cases for plugins sits.

> You mention a lot of interesting points but there are some where we disagree. For example, I see plugins being able to modify the database schema as a problem.

Is the problem the current plugin model or rather allowing your users to shoot themselves in the foot with it?

> There are cases where latency may be a problem and we're investigating WebAssembly as a potential alternative to web hooks for extremely low-latency paths but in general, I believe composition beats code injection on all fronts, maintainability being the biggest.

I see a few frameworks take this path; but for me its really about data consistency. I think the e-commerce domain has strongly sedimented models for Users, Orders, Products and so on. I see it as very hard/impossible for Saleor as a data model to make all clients use-cases happy without allowing them to store data close to those models.

I think Saleor is an amazing piece of software; and the ability to extend it as a plugin model and allow these plugins to store data with strong relations to the Saleor primitives is a great solution since it would extend the graphql model as a result. At the cost of my own code management.

**Reply by PlusA2M** [reply]
We also agree with the point that for our own code (plug-in), we, as the plug-in developer, should take the responsibility to the consequences. Plug-in is a way to let us easily extend the functionality to Saleor with our own way, it's similar to modifying the core directly, though with the plug-in mechanism, we could build it within a better protocol/scope (plug-in), which makes our work a lot easier! In our case, we're producing custom made products, the price calculation model is so complex that even variants would no meet our needs so that we could only build one based on Saleor's core through plug-in. I think the possibility for plug-in is nearly unlimited, that's why we hope it could be kept in whatever way, I don't think Saleor should take any responsibility of any plug-in that introducing bugs or vulnerability. I think it's just like someone modifying the core of the Saleor for their own needs.

**Reply by patrys** [reply]
@PlusA2M Saleor now has an API to add prices to cart with custom price overrides, you may want to look into that 😉 

https://docs.saleor.io/docs/3.x/developer/api-reference/inputs/checkout-line-input#price-positivedecimal

**Reply by PlusA2M** [reply]
@patrys Thanks for the info!
We will look into that soon! Can't wait to see more webhooks release and extending abilities like this! 😎
